Add fee rate distribution in -getinfo

prayank23 opened the pull request:
This PR is on top of https://github.com/bitcoin/bitcoin/pull/21422

Context: https://github.com/bitcoin/bitcoin/issues/17314#issuecomment-890688519

### What?

Add fee rate distribution information in -getinfo

### Why?

1. `-getinfo` has almost everything except a section for mempool and some information which can be helpful while doing transactions or analysis.
2. All PRs trying to add related things in GUI are dead: https://github.com/bitcoin-core/gui/pull/320#issuecomment-912821023. We can try adding basic things in `-getinfo` and see if it helps users. It can be be optional and only work if some argument is used.
3. Fee rate distribution in mempool helps more than any fee estimation because there is nothing to predict. You just need to look at fee rates used by others and bid for block space accordingly.

### How?

Approach similar to progress bar which was added in https://github.com/bitcoin/bitcoin/pull/22547

- Use results from `bitcoin-cli getmempoolinfo "[1,10,100,200,1000]"`
- Use dynamic increment for bar in GetFeerateBar() so that it works for different mempool size and all chains.
- Magenta color for mempool section


### Testing

1. Run `bitcoind` (testnet, signet and mainnet)
2. Run command: `watch -c -t -n 1 bitcoin-cli -getinfo -color=always`
3. Check if everything look okay. Press Ctrl+C. Check different edge cases.

![image](https://github.com/prayank23/bitcoin/raw/gif/mempool-getinfo.gif)

kiminuo committed:
Introduce fee histogram in getmempoolinfo RPC command

Co-authored-by: Jonas Schnelli <dev@jonasschnelli.ch>
Co-authored-by: Jon Atack <jon@atack.com>

kiminuo committed:
test: Add mempool fee histogram test coverage

Original commit: https://github.com/bitcoin/bitcoin/commit/0b6ba66238c377116bc6c21e19cffbf1b6dfc788

Co-authored-by: João Barbosa <joao.paulo.barbosa@gmail.com>
Co-authored-by: Jon Atack <jon@atack.com>

Commit:
add fee_rate_distribution in getinfo

+ 4 fee rate groups: 1-9,10-99,100-199,Above 200
+ Use dynamic increment based on size
+ MAGENTA color for mempool section, CYAN for wallet
+ Do not show mempool section and fee rate group if size is zero
+ Add doxygen documentation for `GetFeerateBar`

Co-authored-by: Klement Tan <klementtan@gmail.com>

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -1027,7 +1053,31 @@ static void ParseGetInfoResult(UniValue& result)
+        break;
+    case 1: fee_rate = "10-19";
+        break;
+    case 2: fee_rate = "20-29";
```
These fee rate ranges seem arbitrary. Relevant real fee rates may evolve to be much higher or lower.

prayank23 replied:
I used them to test. Was planning to add more once results are as expected. We can do one of these to improve:

1. Add more groups: https://i.imgur.com/ffyusBA.png
2. Define a MIN and MAX, MIN will be 1 and MAX can be highest fee rate used by transactions in mempool. And then divide this in two few groups.

prayank23 replied:
Used different groups in last commit: 1-9, 10-99, 100-199

jonatack reviewed (commented):
Various initial thoughts:

- is #21422 enough; what is the added value of an additional CLI?
- if a CLI command would add additional value relative to an RPC, -getinfo is already a bit slow due to calling several RPCs; maybe a separate CLI -feerateinfo or -mempoolinfo command

prayank23 commented:
@jonatack Thanks for review

> is Add feerate histogram to getmempoolinfo #21422 enough; what is the added value of an additional CLI?

I consider `-getinfo` as a dashboard for users who do not like GUI but still prefer to see a summary open in a terminal while using Bitcoin Core for several things. If used with `watch` it can even also update regularly. 

> if a CLI command would add additional value relative to an RPC, -getinfo is already a bit slow due to calling several RPCs; maybe a separate CLI -feerateinfo or -mempoolinfo command

Agree this will add one more RPC in it and things can be slow sometimes. So we can do one of these:

1. Add some option which is `false` by default and `-getinfo` will not return fee rate distribution until user mentions `bitcoin -getinfo mempool=true`
2. Create a new CLI option as you suggested only for this information (-feerateinfo or -mempoolinfo)
3. Add it in RPC `getmempoolinfo` with an argument `feerate_dist` which is false by default

klementtan commented on src/bitcoin-cli.cpp:
```diff
@@ -315,6 +316,7 @@ class GetinfoRequestHandler: public BaseRequestHandler
         result.push_back(JSONRPCRequestObj("getblockchaininfo", NullUniValue, ID_BLOCKCHAININFO));
         result.push_back(JSONRPCRequestObj("getwalletinfo", NullUniValue, ID_WALLETINFO));
         result.push_back(JSONRPCRequestObj("getbalances", NullUniValue, ID_BALANCES));
+        result.push_back(JSONRPCRequestObj("getmempoolinfo", "[1,10,20,30]", ID_MEMPOOLINFO));
```
Updating the code to this solved the error for me.

```suggestion
        UniValue params{RPCConvertValues("generatetoaddress", {"[1,10,20,30]"})};
        result.push_back(JSONRPCRequestObj("getmempoolinfo", params, ID_MEMPOOLINFO));
```

klementtan commented on src/bitcoin-cli.cpp:
```diff
@@ -1027,7 +1053,31 @@ static void ParseGetInfoResult(UniValue& result)
     }
-
+    
+    result_string += strprintf("Feerate Distribution: \n");
```
Maybe add color to standardize with the other headers?

prayank23 replied:
Changed 'Feerate Distribution' to 'Mempool' and added RED color

klementtan reviewed (commented):
Concept ACK

prayank23 commented:
Result for `bitcoin-cli -getinfo` after last commit:

![image](https://user-images.githubusercontent.com/13405205/132280052-ff15d960-c965-4b12-85f4-9f6300e2961f.png)

Changes:

- Change feerate groups
- Increment by 1000
- Use L_BAR for vbytes less than 250k
- Add RED color and use it for MEMPOOL
- Fix fee_size in for lopp
- Only show fee distribution for a group if size>0

@klementtan Thanks for review and help in fixing JSON error

lsilva01 commented:
Concept ACK. It's a really nice feature.
It would be great if the fee rate ranges were automatically calculated based on the minimum and maximum fees.

prayank23 commented:
@lsilva01 Thanks for review. Made few changes in last commit. Let me know if this looks better. Will try few more things in next few days.

Result for `bitcoin-cli -getinfo` after last commit: 

![image](https://user-images.githubusercontent.com/13405205/132926798-ea7fc16c-75a4-4f54-a521-e7c572c1e18f.png)

Changes:

1. Add one more group for fee rates above 200
2. Change colors: 
   Mempool: RED -> MAGENTA
   Wallet: MAGENTA -> CYAN (Wallet and Balance are related so can use the same color. Red color looks weird for anything except errors.)
3. Do not show Mempool section if mempool size is less than zero

TO DO:

- [ ] Dynamic INCREMENT. Use INCREMENT based on minimum size.
- [ ] `mempool=true` argument
- [ ] Other things based on reviews

prayank23 commented:
I have made few changes and squashed commits. Also updated things in PR description. Please test and let me know if there are any issues and scope for improvement. 

@klementtan I tried adding you as co-author because PR is inspired by your work to add progress bar for sync, used lot of code from it and you helped in fixing JSON error initially. Maybe didn't add it correctly. Will have to commit again anyway, so will fix it. Please share your email address that I should use.

Cc: @jonatack @lsilva01

klementtan commented:
@prayank23 My email is klementtan@gmail.com 

Will review the PR soon 👍

klementtan reviewed (commented):
tested ACK 58fd0a4d4431f084f7f8260ed453be0c6d984517 

![image](https://user-images.githubusercontent.com/49265907/133022348-6b40aaaf-9e9c-41ce-ac95-e69624b72032.png)

prayank23 commented on src/bitcoin-cli.cpp:
```diff
@@ -359,6 +362,13 @@ class GetinfoRequestHandler: public BaseRequestHandler
+        result.pushKV("feerate_dist2", batch[ID_MEMPOOLINFO]["result"]["fee_histogram"]["fee_rate_groups"]["10"]["size"]);
+        result.pushKV("feerate_dist3", batch[ID_MEMPOOLINFO]["result"]["fee_histogram"]["fee_rate_groups"]["100"]["size"]);
+        result.pushKV("feerate_dist4", batch[ID_MEMPOOLINFO]["result"]["fee_histogram"]["fee_rate_groups"]["200"]["size"]);
+        result.pushKV("mempool_bytes", batch[ID_MEMPOOLINFO]["result"]["bytes"]);
```
Yes. Because fee rate groups are:

1 (from 1 to 9)
10 (from 10 to 99)
100 (from 100 to 199)
200 (from 200 to 1000)

You can see them in the results for `bitcoin-cli getmempoolinfo "[1,10,100,200,1000]"`

klementtan replied:
Ah got it! Tx with fee >= 1000 sat/vB will not be displayed.

Would it be better to update s/Above 200  |/200-1000   |/? IIUC `batch[ID_MEMPOOLINFO]["result"]["fee_histogram"]["fee_rate_groups"]["200"]["size"]` only considers fees from `200-1000`

```
./src/bitcoin-cli -signet getmempoolinfo "[1,10,100,200,1000]"
...
      "200": {
        "size": 0,
        "count": 0,
        "fees": 0,
        "from": 200,
        "to": 999
      },
...
```

prayank23 replied:
Possible. Or we can change 1000 to a higher number. Will see what other reviewers think and decide accordingly.

prayank23 replied:
Will keep this unchanged as nobody reviewed and had issues with it. Fee rates are normally less than 1000 sat/vB and we can ignore if someone uses higher in few transactions because this is just the fee rate distribution summary to help users and complete information.

prayank23 commented on src/bitcoin-cli.cpp:
```diff
@@ -1028,6 +1073,64 @@ static void ParseGetInfoResult(UniValue& result)
         result_string += "\n";
     }
 
+    const double mempool_bytes{result["mempool_bytes"].get_real()};
```
A function for whole thing? Yes.

klementtan replied:
Yup L1076-L1130 looks like it could be cleanly refactored into another function 👍

prayank23 replied:
Sorry Its been more than a month. I was waiting for some progress in PR 21422 before making any changes here as it depends on it. Anyways looked at the code today and this looks alright because everything is inside `ParseGetInfoResult()`

prayank23 commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,39 @@ static void GetProgressBar(double progress, std::string& progress_bar)
 }
 
+
+static void GetFeerateBar(double size, std::string& size_bar, double inc_size, double max_size)
```
Sure

prayank23 replied:
Done

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,46 @@ static void GetProgressBar(double progress, std::string& progress_bar)
+ *
+ * @param[in]  size      Size of fee rate group in bytes.
+ * @param[out] size_bar  String representation of the distribution bar.
+ * @param[in]  inc_size  Minimum size of fee rate groups in bytes.
```
Is the comment correct?

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,46 @@ static void GetProgressBar(double progress, std::string& progress_bar)
+/**
+ * GetFeerateBar constructs fee rate distribution bars using dynamic increment based on size.
+ *
+ * @param[in]  size      Size of fee rate group in bytes.
```
Maybe the arguments can be re-ordered so that `in` params are first and the `out` param is last.

prayank23 replied:
Done

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,46 @@ static void GetProgressBar(double progress, std::string& progress_bar)
+    static double INCREMENT{inc_size};
+    static const std::string BAR{"\u2591"};
+
+    if (max_size < 10000){
```
Given that there so many very similar numbers differing in the number of zero digits, one may consider using `10'000` syntax to represent `10000` (https://en.cppreference.com/w/cpp/language/integer_literal). (I'm not sure if it is allowed in this repo though)

prayank23 replied:
Not sure about this as I have not seen such format for any constants in this repository.

kiminuo replied:
IDK, maybe @jonatack would know. 🙏

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,46 @@ static void GetProgressBar(double progress, std::string& progress_bar)
+    	for (int i = 0; i < size / (INCREMENT*1000); ++i) {
+            size_bar += BAR;
+        }
+    } else if (max_size > 10000000 && max_size < 100000000){
```
```suggestion
    } else if (max_size > 10000000 && max_size < 100000000) {
```

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,46 @@ static void GetProgressBar(double progress, std::string& progress_bar)
+        for (int i = 0; i < size / INCREMENT; ++i) {
+            size_bar += BAR;
+        }
+    } else if (max_size > 10000 && max_size < 100000){
```
What will happen for value `100000`?

prayank23 replied:
Good catch. Will have to make some changes.

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -928,6 +938,46 @@ static void GetProgressBar(double progress, std::string& progress_bar)
+            size_bar += BAR;
+        }
+    } else if (max_size > 100000 && max_size < 1000000){
+    	for (int i = 0; i < size / (INCREMENT*100); ++i) {
+            size_bar += BAR;
+        }
```
All the for-cycles are the same basically, so maybe you can have one for-cycle after this set of conditions. That may actually be easier to review.

prayank23 replied:
Right this can be done. Will do later today.

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -1028,6 +1080,64 @@ static void ParseGetInfoResult(UniValue& result)
+
+    	//Removing zero from numbers used in std::min to avoid division by zero in GetFeerateBar()
+    	auto const ignore_zero = [](auto const& a, auto const& b) -> bool {
+        	if(0 == a || 0 == b)
```
```suggestion
        	if (0 == a || 0 == b)
```

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -1028,6 +1080,64 @@ static void ParseGetInfoResult(UniValue& result)
+    	};
+
+    	double max_size = std::max({size1, size2, size3, size4});
+    	double inc_size = std::min({size1, size2, size3, size4}, ignore_zero);
```
What should happen when all values are zero?

prayank23 replied:
`mempool_bytes` will be zero in that case and mempool section will not be printed L1085

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -1028,6 +1080,64 @@ static void ParseGetInfoResult(UniValue& result)
+    		case 3: fee_rate = "100-199   |";
+    		        fee_size = size3;
+        		break;
+        	case 4: fee_rate = "Above 200 |";
```
This line makes me think that value `200` is not included neither in case 3 nor in case 4.

prayank23 replied:
200 is included in case 4. I wrote `Above 200` in short for `200 and Above 200`

kiminuo replied:
Yeah. Maybe one can use `≥ 200` (https://www.fileformat.info/info/unicode/char/2265/index.htm) instead of "Above 200" to make the meaning absolutely clear.

kiminuo reviewed (commented):
Just a few comments based on reading the code

nopara73 commented:
Feature ACK

kiminuo commented on src/bitcoin-cli.cpp:
```diff
@@ -315,6 +316,8 @@ class GetinfoRequestHandler: public BaseRequestHandler
         result.push_back(JSONRPCRequestObj("getblockchaininfo", NullUniValue, ID_BLOCKCHAININFO));
         result.push_back(JSONRPCRequestObj("getwalletinfo", NullUniValue, ID_WALLETINFO));
         result.push_back(JSONRPCRequestObj("getbalances", NullUniValue, ID_BALANCES));
+        UniValue params{RPCConvertValues("getmempoolinfo", {"[1,10,100,200,1000]"})};
```
Playing devil's advocate: Why to use these values and not different ones? 

I think it's very hard to argue that these values are the best possible. Without more information in PR's description, this feels arbitrary. And even if you can persuade people for Bitcoin mainnet chain, I would be surprised if these values were good for Bitcoin testnet chain.

I don't really know how to approach this but it seems important. This is also the reason why #21422 lets user to decide. Original #15836 [mentioned log scale](https://github.com/bitcoin/bitcoin/pull/15836/files#diff-decae4be02fb8a47ab4557fe74a9cb853bdfa3ec0fa1b515c0a1e5de91f4ad0bR1497) which seems in principle much better to me.

prayank23 replied:
> Playing devil's advocate: Why to use these values and not different ones?

Different can be either random values which is difficult to support in this feature which is just mean to be a summary of mempool. Or use more values in between which will use more space hence used all multiples of 10 until 100 and then trying to include everything else in 'Above 200' category.

> And even if you can persuade people for Bitcoin mainnet chain, I would be surprised if these values were good for Bitcoin testnet chain.

I tied them on testnet and mainnet. Works fine because it wont print a category if it has zero size.

> This is also the reason why #21422 lets user to decide.

This is not replacement for 21422 but a usecase for it.

> Original #15836 mentioned log scale which seems in principle much better to me.

It was closed by one of the maintainers with comment: "Closing this, given it's been taken over in #21422"


![image](https://user-images.githubusercontent.com/13405205/141219327-18086649-6c4a-43df-b208-ca849efa9f72.png)

kiminuo replied:
> > And even if you can persuade people for Bitcoin mainnet chain, I would be surprised if these values were good for Bitcoin testnet chain.
> 
> I tried them on testnet and mainnet. Works fine because it wont print a category if it has zero size.

I'm not sure we understand each other. Anyway, I just wanted to point out that any fixed intervals may sooner or later become obsolete. Suppose that your PR is merged and for whatever reasons nobody will mine fee rates less than 200. Then you will just show one bar (right?) and that's not useful really. That is my main concern and that's why I mentioned #15836 and and the log scale (a nice curve with interesting mathematical properties) - again, I'm not saying "use this", just trying to say that some math analysis might be needed here.

> > This is also the reason why #21422 lets user to decide.
> 
> This is not replacement for 21422 but a usecase for it.

Yes, I know.

> One example of fee rate distribution from https://btc.bitaps.com/:
> 
> ![image](https://user-images.githubusercontent.com/13405205/141219327-18086649-6c4a-43df-b208-ca849efa9f72.png)

I like it but it's a website and they can change the output whenever they like. I don't consider this to be true for Bitcoin Core.

Commit:
Update src/bitcoin-cli.cpp

Co-authored-by: kiminuo <58662979+kiminuo@users.noreply.github.com>

prayank23 commented on src/bitcoin-cli.cpp:
```diff
@@ -1028,6 +1080,64 @@ static void ParseGetInfoResult(UniValue& result)
     }
 
+    const double mempool_bytes{result["mempool_bytes"].get_real()};
+
```
I think I used different editor while working on this and my VS Code is behaving weird since few days. Will try to fix this.

Commit:
Update src/bitcoin-cli.cpp

Co-authored-by: kiminuo <58662979+kiminuo@users.noreply.github.com>

Commit:
Update src/bitcoin-cli.cpp

Co-authored-by: kiminuo <58662979+kiminuo@users.noreply.github.com>

Commit:
reorder params in comments

Commit:
Update src/bitcoin-cli.cpp

Co-authored-by: kiminuo <58662979+kiminuo@users.noreply.github.com>

Commit:
Update src/bitcoin-cli.cpp

Co-authored-by: kiminuo <58662979+kiminuo@users.noreply.github.com>
